### DISCUSSION TITLE
posts/understanding_the_math_behind_restir_di/

### DISCUSSION THREAD

**Discussion by giscus** [JiayinCao/AGraphicsGuyNotes#12] [Giscus-Comments]
# posts/understanding_the_math_behind_restir_di/

Ever since the introduction of RTX technology, the industry pushed really hard on real time GI solutions. ReStir (Reservoir Spatio-Temporal Importance Resampling) is one of the new popular topics lately. The algorithm can be applied in multiple applications, there are several different variations of the algorithm, like ReStir DI, ReStir GI, ReStir PT, Volumetric ReStir, etc.
The original papers are clearly the best resources to learn the tech. Due to the complexity of the algorithm, understanding the theory and math remains challenging for graphics programmers who are not very familiar with sampling methods.

https://agraphicsguynotes.com/posts/understanding_the_math_behind_restir_di/

**Comment by Rainxyh** [comment]
Respect.

**Comment by changxy2** [comment]
nice

**Comment by Mephisto405** [comment]
![image](https://github.com/JiayinCao/AGraphicsGuyNotes/assets/25499931/32990472-f292-41d3-9e1c-a18883ba1126)
Hello. Your post is truly a gift in itself. It was very difficult to integrate the scattered explanations in various places, but many things became easily understood after reading your post. However, I think there might be a typo in the image above, am I correct?

**Reply by Mephisto405** [reply]
According to the other papers:
Decorrelating ReSTIR Samplers via MCMC Mutations
![image](https://github.com/JiayinCao/AGraphicsGuyNotes/assets/25499931/799345a3-20c8-4d2d-a4e5-0334e90ea6e7)
Spatiotemporal reservoir resampling
![image](https://github.com/JiayinCao/AGraphicsGuyNotes/assets/25499931/b472eee8-9470-4377-8bc5-a48e87e62f58)

**Reply by JiayinCao** [reply]
I see what you mean.
This is already adjusted. Thanks for the input.

**Comment by fstrugar** [comment]
Thank you so much for the great article, it helped me understand a number of resampling things!

Question on WRS sample picking corner cases:

> If we would like to pick it when it is equal, there is a non-zero chance to pick incoming samples with zero weight.
> If we would like to ignore it when it is equal, there is a non-zero chance that we would ignore the first sample, which should have always been picked. Of course, if the first sample has zero weight, it is even trickier.

Isn't this whole thing a non-issue if the random numbers are guaranteed to be in [0, 1) range? Using just `less than threshold` should then cover all cases?

**Reply by JiayinCao** [reply]
I believe you could be right. At least I don't see a major problem with the solution.

I'll adjust this part when I find time later.

**Comment by malikarjun** [comment]
Maybe I am missing something, but in **Commonly Used MIS Weight** section for Uniform Weight should it be $m_k(x) = \frac{1}{M}$ since $m_k(x) = \frac{1}{N_k}$ will not sum up to 1 for all $k$ in $[1, M]$?

**Reply by JiayinCao** [reply]
You are right about that I made a mistake there.
But I'm afraid that your suggestion is not mathematically unbiased either.
I have updated the math. Let me know if there is further confusions.

Thanks for pointing it out to me.
